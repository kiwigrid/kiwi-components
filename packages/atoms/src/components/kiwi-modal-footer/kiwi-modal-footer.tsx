import {
  Component,
  Host,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  Fragment,
} from '@stencil/core';

/**
 * @slot kiwi-modal-footer-cancel - The content for the default cancel button.
 * @slot kiwi-modal-footer-confirm - The content for the default confirm button.
 */
@Component({
  tag: 'kiwi-modal-footer',
  shadow: false,
})
export class KiwiModalFooter {
  /**
   * The labels to be used for the default footer buttons, passed as a tuple of
   * strings with the first entry being the confirm label, and the second being
   * the cancel label.
   */
  @Prop()
  defaultLabels: [confirm: string, cancel: string] = ['Ok', 'Cancel'];

  /**
   * Whether to render the default footer buttons or a slot.
   *
   * This is a workaround since slot fallback content is not rendered correctly
   * with `shadow: false`
   */
  @Prop()
  useDefault = true;

  /** The event emitted when the user clicks the default cancel button. */
  @Event()
  cancel!: EventEmitter;

  /** The event emitted when the user clicks the default confirm button. */
  @Event()
  confirm!: EventEmitter;

  render(): JSX.Element {
    const [confirmLabel, cancelLabel] = this.defaultLabels;

    return (
      <Host slot="kiwi-modal-footer">
        <div class="modal-footer">
          {this.useDefault ? (
            <Fragment>
              <button class="btn btn-link" onClick={this.handleClose}>
                <slot name="kiwi-modal-footer-cancel">{cancelLabel}</slot>
              </button>
              <button class="btn btn-primary" onClick={this.handleConfirmation}>
                <slot name="kiwi-modal-footer-confirm">{confirmLabel}</slot>
              </button>
            </Fragment>
          ) : (
            <slot></slot>
          )}
        </div>
      </Host>
    );
  }

  private handleClose = (): void => {
    this.cancel.emit();
  };

  private handleConfirmation = (): void => {
    this.confirm.emit();
  };
}
