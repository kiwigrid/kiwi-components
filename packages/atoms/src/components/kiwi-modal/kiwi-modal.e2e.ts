import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-modal', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-modal></kiwi-modal>');

    const element = await page.find('kiwi-modal');
    expect(element).toHaveClass('hydrated');
  });
});
