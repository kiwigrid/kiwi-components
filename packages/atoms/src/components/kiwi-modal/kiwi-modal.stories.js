import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Modal',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

const openModal = () => (document.querySelector('kiwi-modal').open = true);

export const basic = () => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>
    <kiwi-modal>
      <div slot="modal-body">
        There is just this text and nothing else. Close me by clicking outside.
      </div>
    </kiwi-modal>`;
};

export const full = () => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>
    <kiwi-modal with-header cancel-text="Cancel" ok-text="Ok" escape>
      <i slot="modal-title">Italic Modal Header</i>
      <div slot="modal-body">Press Escape to close the modal</div>
    </kiwi-modal>`;
};

export const stepped = () => {
  const setStep1 = (event) => {
    if (event.target.id === 'steppedModal') {
      event.target.nextText = 'Next';
      event.target.cancelText = 'Cancel';
      event.target.previousText = undefined;
      event.target.okText = undefined;
      event.target.querySelector('[slot="modal-body"]').textContent = 'Step 1';
    }
  };
  const setStep2 = (event) => {
    if (event.target.id === 'steppedModal') {
      event.target.nextText = undefined;
      event.target.cancelText = 'Cancel';
      event.target.previousText = 'Back';
      event.target.okText = 'Finish';
      event.target.querySelector('[slot="modal-body"]').textContent = 'Step 2';
    }
  };
  document.addEventListener('next', setStep2);
  document.addEventListener('previous', setStep1);
  document.addEventListener('closed', setStep1);

  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>
          Show Stepped Modal
        </button>
      </div>
    </div>
    <kiwi-modal
      id="steppedModal"
      with-header="true"
      cancel-text="Cancel"
      next-text="Next"
    >
      <i slot="modal-title">Italic Modal Header</i>
      <div slot="modal-body">Step 1</div>
    </kiwi-modal>`;
};

export const scrolling = () => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
        <p style="width: 100px">
          Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam
          nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat,
          sed diam voluptua. At vero eos et accusam et justo duo dolores et ea
          rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem
          ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur
          sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et
          dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam
          et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea
          takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit
          amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
          invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua.
          At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
          kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit
          amet. Duis autem vel eum iriure dolor in hendrerit in vulputate velit
          esse molestie consequat, vel illum dolore eu feugiat nulla facilisis
          at vero eros et accumsan et iusto odio dignissim qui blandit praesent
          luptatum zzril delenit augue duis dolore te feugait nulla facilisi.
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam
          nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat
          volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation
          ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat.
          Duis autem vel eum iriure dolor in hendrerit in vulputate velit esse
          molestie consequat, vel illum dolore eu feugiat nulla facilisis at
          vero eros et accumsan et iusto odio dignissim qui blandit praesent
          luptatum zzril delenit augue duis dolore te feugait nulla facilisi.
          Nam liber tempor cum soluta nobis eleifend option congue nihil
          imperdiet doming id quod mazim placerat facer possim assum. Lorem
          ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy
          nibh euismod tincidunt ut laoreet dolore magna aliquam erat volutpat.
          Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper
          suscipit lobortis nisl ut aliquip ex ea commodo consequat. Duis autem
          vel eum iriure dolor in hendrerit in vulputate velit esse molestie
          consequat, vel illum dolore eu feugiat nulla facilisis. At vero eos et
          accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren,
          no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum
          dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
          tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
          voluptua. At vero eos et accusam et justo duo dolores et ea rebum.
          Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum
          dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing
          elitr, At accusam aliquyam diam diam dolore dolores duo eirmod eos
          erat, et nonumy sed tempor et et invidunt justo labore Stet clita ea
          et gubergren, kasd magna no rebum. sanctus sea sed takimata ut vero
          voluptua. est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet,
          consetetur
        </p>
      </div>
    </div>
    <kiwi-modal with-header escape>
      <div slot="modal-body">
        This is a very long text to show and test scrolling behaviour. Lorem
        ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
        eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
        voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet
        clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit
        amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam
        nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat,
        sed diam voluptua. At vero eos et accusam et justo duo dolores et ea
        rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem
        ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing
        elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna
        aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo
        dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus
        est Lorem ipsum dolor sit amet. Duis autem vel eum iriure dolor in
        hendrerit in vulputate velit esse molestie consequat, vel illum dolore
        eu feugiat nulla facilisis at vero eros et accumsan et iusto odio
        dignissim qui blandit praesent luptatum zzril delenit augue duis dolore
        te feugait nulla facilisi. Lorem ipsum dolor sit amet, consectetuer
        adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet
        dolore magna aliquam erat volutpat. Ut wisi enim ad minim veniam, quis
        nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex
        ea commodo consequat. Duis autem vel eum iriure dolor in hendrerit in
        vulputate velit esse molestie consequat, vel illum dolore eu feugiat
        nulla facilisis at vero eros et accumsan et iusto odio dignissim qui
        blandit praesent luptatum zzril delenit augue duis dolore te feugait
        nulla facilisi. Nam liber tempor cum soluta nobis eleifend option congue
        nihil imperdiet doming id quod mazim placerat facer possim assum. Lorem
        ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy
        nibh euismod tincidunt ut laoreet dolore magna aliquam erat volutpat. Ut
        wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper
        suscipit lobortis nisl ut aliquip ex ea commodo consequat. Duis autem
        vel eum iriure dolor in hendrerit in vulputate velit esse molestie
        consequat, vel illum dolore eu feugiat nulla facilisis. At vero eos et
        accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no
        sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor
        sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
        invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
        vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
        gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
        ipsum dolor sit amet, consetetur sadipscing elitr, At accusam aliquyam
        diam diam dolore dolores duo eirmod eos erat, et nonumy sed tempor et et
        invidunt justo labore Stet clita ea et gubergren, kasd magna no rebum.
        sanctus sea sed takimata ut vero voluptua. est Lorem ipsum dolor sit
        amet. Lorem ipsum dolor sit amet, consetetur
      </div>
    </kiwi-modal>`;
};
