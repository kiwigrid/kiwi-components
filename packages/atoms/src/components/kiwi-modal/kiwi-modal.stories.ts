import { html, TemplateResult } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Modal',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

const toggleModal = (open: boolean) => (): void => {
  const modal = document.querySelector('kiwi-modal');

  if (!modal) {
    return;
  }

  modal.open = open;
};

const openModal = toggleModal(true);
const closeModal = toggleModal(false);

export const basic = (): TemplateResult => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>
    <kiwi-modal>
      <div slot="kiwi-modal-body">
        There is just this text and nothing else. Close me by clicking outside.
      </div>
    </kiwi-modal>`;
};

export const fullWithDefaultFooter = (): TemplateResult => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>

    <kiwi-modal with-header escape>
      <i slot="kiwi-modal-title">Italic Modal Header</i>

      <div slot="kiwi-modal-body">Press Escape to close the modal</div>

      <kiwi-modal-footer
        slot="kiwi-modal-footer"
        @cancel=${closeModal}
        @confirm=${closeModal}
      ></kiwi-modal-footer>
    </kiwi-modal>`;
};

export const fullWithCustomFooterLabels = (): TemplateResult => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>

    <kiwi-modal with-header escape>
      <i slot="kiwi-modal-title">Italic Modal Header</i>

      <div slot="kiwi-modal-body">Press Escape to close the modal</div>

      <kiwi-modal-footer
        .defaultLabels=${['Confirm', 'Abort']}
        @cancel=${closeModal}
        @confirm=${closeModal}
        slot="kiwi-modal-footer"
      ></kiwi-modal-footer>
    </kiwi-modal>`;
};

export const fullWithCustomFooterButtonContents = (): TemplateResult => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>

    <kiwi-modal with-header escape>
      <i slot="kiwi-modal-title">Italic Modal Header</i>

      <div slot="kiwi-modal-body">Press Escape to close the modal</div>

      <kiwi-modal-footer
        slot="kiwi-modal-footer"
        @cancel=${closeModal}
        @confirm=${closeModal}
      >
        <span slot="kiwi-modal-footer-confirm">
          <span class="glyphicon glyphicon-ok mr-1"></span>
          Okay
        </span>

        <span slot="kiwi-modal-footer-cancel">
          <span class="glyphicon glyphicon-remove mr-1"></span>
          Abbrechen
        </span>
      </kiwi-modal-footer>
    </kiwi-modal>`;
};

export const fullWithCustomFooter = (): TemplateResult => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>
      </div>
    </div>

    <kiwi-modal with-header escape>
      <i slot="kiwi-modal-title">Italic Modal Header</i>

      <div slot="kiwi-modal-body">Press Escape to close the modal</div>

      <kiwi-modal-footer .useDefault=${false} slot="kiwi-modal-footer">
        <button class="btn btn-danger" @click=${closeModal}>Delete</button>
      </kiwi-modal-footer>
    </kiwi-modal>`;
};

export const scrolling = (): TemplateResult => {
  return html`<div class="panel panel-default m-1">
      <div class="panel-body">
        <button class="btn btn-default" @click=${openModal}>Show Modal</button>

        <hr />

        <h2>This is long content</h2>

        <p style="width: 100px">
          Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam
          nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat,
          sed diam voluptua. At vero eos et accusam et justo duo dolores et ea
          rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem
          ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur
          sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et
          dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam
          et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea
          takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit
          amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
          invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua.
          At vero eos et accusam et justo duo dolores et ea rebum. Stet clita
          kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit
          amet. Duis autem vel eum iriure dolor in hendrerit in vulputate velit
          esse molestie consequat, vel illum dolore eu feugiat nulla facilisis
          at vero eros et accumsan et iusto odio dignissim qui blandit praesent
          luptatum zzril delenit augue duis dolore te feugait nulla facilisi.
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit, sed diam
          nonummy nibh euismod tincidunt ut laoreet dolore magna aliquam erat
          volutpat. Ut wisi enim ad minim veniam, quis nostrud exerci tation
          ullamcorper suscipit lobortis nisl ut aliquip ex ea commodo consequat.
          Duis autem vel eum iriure dolor in hendrerit in vulputate velit esse
          molestie consequat, vel illum dolore eu feugiat nulla facilisis at
          vero eros et accumsan et iusto odio dignissim qui blandit praesent
          luptatum zzril delenit augue duis dolore te feugait nulla facilisi.
          Nam liber tempor cum soluta nobis eleifend option congue nihil
          imperdiet doming id quod mazim placerat facer possim assum. Lorem
          ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy
          nibh euismod tincidunt ut laoreet dolore magna aliquam erat volutpat.
          Ut wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper
          suscipit lobortis nisl ut aliquip ex ea commodo consequat. Duis autem
          vel eum iriure dolor in hendrerit in vulputate velit esse molestie
          consequat, vel illum dolore eu feugiat nulla facilisis. At vero eos et
          accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren,
          no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum
          dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod
          tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
          voluptua. At vero eos et accusam et justo duo dolores et ea rebum.
          Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum
          dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing
          elitr, At accusam aliquyam diam diam dolore dolores duo eirmod eos
          erat, et nonumy sed tempor et et invidunt justo labore Stet clita ea
          et gubergren, kasd magna no rebum. sanctus sea sed takimata ut vero
          voluptua. est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet,
          consetetur
        </p>
      </div>
    </div>

    <kiwi-modal escape>
      <div slot="kiwi-modal-body">
        This is a very long text to show and test scrolling behaviour. Lorem
        ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy
        eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam
        voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet
        clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit
        amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam
        nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat,
        sed diam voluptua. At vero eos et accusam et justo duo dolores et ea
        rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem
        ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing
        elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna
        aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo
        dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus
        est Lorem ipsum dolor sit amet. Duis autem vel eum iriure dolor in
        hendrerit in vulputate velit esse molestie consequat, vel illum dolore
        eu feugiat nulla facilisis at vero eros et accumsan et iusto odio
        dignissim qui blandit praesent luptatum zzril delenit augue duis dolore
        te feugait nulla facilisi. Lorem ipsum dolor sit amet, consectetuer
        adipiscing elit, sed diam nonummy nibh euismod tincidunt ut laoreet
        dolore magna aliquam erat volutpat. Ut wisi enim ad minim veniam, quis
        nostrud exerci tation ullamcorper suscipit lobortis nisl ut aliquip ex
        ea commodo consequat. Duis autem vel eum iriure dolor in hendrerit in
        vulputate velit esse molestie consequat, vel illum dolore eu feugiat
        nulla facilisis at vero eros et accumsan et iusto odio dignissim qui
        blandit praesent luptatum zzril delenit augue duis dolore te feugait
        nulla facilisi. Nam liber tempor cum soluta nobis eleifend option congue
        nihil imperdiet doming id quod mazim placerat facer possim assum. Lorem
        ipsum dolor sit amet, consectetuer adipiscing elit, sed diam nonummy
        nibh euismod tincidunt ut laoreet dolore magna aliquam erat volutpat. Ut
        wisi enim ad minim veniam, quis nostrud exerci tation ullamcorper
        suscipit lobortis nisl ut aliquip ex ea commodo consequat. Duis autem
        vel eum iriure dolor in hendrerit in vulputate velit esse molestie
        consequat, vel illum dolore eu feugiat nulla facilisis. At vero eos et
        accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no
        sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor
        sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor
        invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At
        vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd
        gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem
        ipsum dolor sit amet, consetetur sadipscing elitr, At accusam aliquyam
        diam diam dolore dolores duo eirmod eos erat, et nonumy sed tempor et et
        invidunt justo labore Stet clita ea et gubergren, kasd magna no rebum.
        sanctus sea sed takimata ut vero voluptua. est Lorem ipsum dolor sit
        amet. Lorem ipsum dolor sit amet, consetetur
      </div>
    </kiwi-modal>`;
};
