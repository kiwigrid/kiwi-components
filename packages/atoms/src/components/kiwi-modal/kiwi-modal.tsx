import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  Listen,
  Prop,
  Watch,
} from '@stencil/core';
import { some } from 'lodash-es';

export type Size = 'small' | 'medium' | 'large';
const sizeClasses: { [Key in Size]: string } = {
  small: 'modal-sm',
  medium: '',
  large: 'modal-lg',
};

@Component({
  tag: 'kiwi-modal',
  shadow: false,
})
export class KiwiModal {
  /** Set this to true to show the modal or alternatively set an id to this element and dispatch a 'showKiwiModal' CustomEvent with the id as event.detail */
  @Prop({
    mutable: true,
    reflect: true,
  })
  open = false;

  /** Set this to true if you want to show the header */
  @Prop()
  withHeader = false;

  /** Set this text to show the cancel button */
  @Prop()
  cancelText?: string;

  /** Set this text to show the previous button */
  @Prop()
  previousText?: string;

  /** Set this text to show the next button */
  @Prop()
  nextText?: string;

  /** Set this text to show the ok button */
  @Prop()
  okText?: string;

  /** Set to true if the modal should be closed on Escape press */
  @Prop()
  escape = false;

  /** Set the size of the modal */
  @Prop()
  size: Size = 'medium';

  @Element()
  host!: HTMLKiwiModalElement;

  /** This event is emitted after the modal was closed */
  @Event()
  closed!: EventEmitter;

  /** This event is emitted on click on the "previous" button  */
  @Event()
  previous!: EventEmitter;

  /** This event is emitted on click on the "next" button */
  @Event()
  next!: EventEmitter;

  /** This event is emitted on click on the "ok" button  */
  @Event()
  confirmed!: EventEmitter;

  private modalElement!: HTMLDivElement;

  private backdropElement!: HTMLDivElement;

  @Listen('showKiwiModal', { target: 'document' })
  private showKiwiModalHandler(event: CustomEvent<string>): void {
    if (event.detail === this.host.id) {
      this.open = true;
    }
  }

  @Watch('open')
  private onOpenChange(newValue: boolean): void {
    if (newValue) {
      this.handleOpeningCssClasses();
    } else {
      this.handleClosingCssClasses();
    }
  }

  componentDidLoad(): void {
    if (this.escape) {
      document.addEventListener('keydown', (e) => {
        if (e.key === 'Escape') {
          this.handleClose();
        }
      });
    }

    this.onOpenChange(this.open);
  }

  disconnectedCallback(): void {
    this.handleClosingCssClasses();
  }

  private handleOpeningCssClasses = (): void => {
    document.body.classList.add('modal-open');
    this.modalElement.classList.add('show', 'fade');
    this.backdropElement.classList.add('show', 'fade');
    window.setTimeout(() => {
      this.modalElement.classList.add('in');
      this.backdropElement.classList.add('in');
    }, 150);
  };

  private handleClosingCssClasses = (): void => {
    this.modalElement.classList.remove('in');
    window.setTimeout(() => {
      this.modalElement.classList.remove('show', 'fade');
      this.backdropElement.classList.remove('in');
    }, 250);
    window.setTimeout(() => {
      this.backdropElement.classList.remove('show', 'fade');
      document.body.classList.remove('modal-open');
    }, 500);
  };

  private handleClose = (): void => {
    this.open = false;
    this.closed.emit();
  };
  private handlePrevious = (): void => {
    this.previous.emit();
  };
  private handleNext = (): void => {
    this.next.emit();
  };
  private handleConfirmation = (): void => {
    this.confirmed.emit();
    this.handleClose();
  };
  private handleClickOutside = (event: Event): void => {
    if (event.target === this.modalElement) {
      this.handleClose();
    }
  };

  render(): void {
    return (
      <Host>
        <div
          class="modal"
          onClick={this.handleClickOutside}
          ref={(el) => (this.modalElement = el as HTMLDivElement)}
        >
          <div class={`modal-dialog ${sizeClasses[this.size]}`}>
            <div class="modal-content">
              {this.withHeader && (
                <div class="modal-header modal-default">
                  <button
                    class="close"
                    type="close"
                    aria-hidden="true"
                    onClick={this.handleClose}
                  >
                    Ã—
                  </button>
                  <h4 class="modal-title">
                    <slot name="modal-title"></slot>
                  </h4>
                </div>
              )}
              <div class="modal-body">
                <slot name="modal-body"></slot>
              </div>
              {some([
                this.cancelText,
                this.previousText,
                this.nextText,
                this.okText,
              ]) && (
                <div class="modal-footer">
                  {this.previousText && (
                    <button
                      class="btn btn-link pull-left"
                      onClick={this.handlePrevious}
                    >
                      {this.previousText}
                    </button>
                  )}
                  {this.cancelText && (
                    <button class="btn btn-link" onClick={this.handleClose}>
                      {this.cancelText}
                    </button>
                  )}
                  {this.nextText && (
                    <button class="btn btn-primary" onClick={this.handleNext}>
                      {this.nextText}
                    </button>
                  )}
                  {this.okText && (
                    <button
                      class="btn btn-primary"
                      onClick={this.handleConfirmation}
                    >
                      {this.okText}
                    </button>
                  )}
                </div>
              )}
            </div>
          </div>
        </div>
        <div
          class="modal-backdrop"
          style={{ display: 'none' }}
          ref={(el) => (this.backdropElement = el as HTMLDivElement)}
        />
      </Host>
    );
  }
}
