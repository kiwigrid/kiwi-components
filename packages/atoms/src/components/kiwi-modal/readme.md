# kiwi-modal

To show the modal there are two options:

- set the attribute `open` of the component
- dispatch a CustomEvent with name `showKiwiModal` and payload `{detail: ${modalId}}` where `modalId` is the id that you gave the component

<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description                                                                                                                                          | Type                             | Default    |
| ------------ | ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------- | ---------- |
| `escape`     | `escape`      | Set to true if the modal should be closed on Escape press                                                                                            | `boolean`                        | `false`    |
| `open`       | `open`        | Set this to true to show the modal or alternatively set an id to this element and dispatch a 'showKiwiModal' CustomEvent with the id as event.detail | `boolean`                        | `false`    |
| `size`       | `size`        | Set the size of the modal                                                                                                                            | `"large" \| "medium" \| "small"` | `'medium'` |
| `withHeader` | `with-header` | Set this to true if you want to show the header                                                                                                      | `boolean`                        | `false`    |


## Events

| Event   | Description                                      | Type               |
| ------- | ------------------------------------------------ | ------------------ |
| `close` | This event is emitted after the modal was closed | `CustomEvent<any>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
