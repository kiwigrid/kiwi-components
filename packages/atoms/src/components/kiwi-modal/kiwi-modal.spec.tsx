import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { KiwiModalFooter } from '../kiwi-modal-footer/kiwi-modal-footer';
import { expectDefined } from '../util/testing';
import { KiwiModal } from './kiwi-modal';

describe('kiwi-modal', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [KiwiModal, KiwiModalFooter],
      template: () => (
        <kiwi-modal id="modalId" withHeader escape>
          <i slot="kiwi-modal-title">Italic Modal Header</i>
          <div slot="kiwi-modal-body">Press Escape to close the modal</div>
          <kiwi-modal-footer slot="kiwi-modal-footer" />
        </kiwi-modal>
      ),
    });

    expectDefined(page.root);
    page.root.open = true;
    await page.waitForChanges();

    expect(page.root).toHaveAttribute('open');
    expect(page.root).toMatchSnapshot();
  });

  it.each`
    size
    ${'small'}
    ${'large'}
  `('renders with size $size', async ({ size }) => {
    const page = await newSpecPage({
      components: [KiwiModal],
      template: () => (
        <kiwi-modal size={size}>
          <i slot="kiwi-modal-title">Italic Modal Header</i>
          <div slot="kiwi-modal-body">{size}</div>
        </kiwi-modal>
      ),
    });

    expect(page.root).toMatchSnapshot();
  });
});
