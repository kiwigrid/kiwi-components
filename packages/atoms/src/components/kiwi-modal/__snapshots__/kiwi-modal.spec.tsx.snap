// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-modal renders 1`] = `
<kiwi-modal cancel-text="Cancel" id="modalId" ok-text="Ok" open="" with-footer="true" with-header="true">
  <!---->
  <div class="fade modal show">
    <div class="modal-dialog">
      <div class="modal-content">
        <div class="modal-default modal-header">
          <button aria-hidden="true" class="close" type="close">
            Ã—
          </button>
          <h4 class="modal-title">
            <i slot="modal-title">
              Italic Modal Header
            </i>
          </h4>
        </div>
        <div class="modal-body">
          <div slot="modal-body">
            Press Escape to close the modal
          </div>
        </div>
        <div class="modal-footer">
          <button class="btn btn-link">
            Cancel
          </button>
          <button class="btn btn-primary">
            Ok
          </button>
        </div>
      </div>
    </div>
  </div>
  <div class="fade modal-backdrop show" style="display: none;"></div>
</kiwi-modal>
`;
